From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shawn Lin <shawn.lin@rock-chips.com>
Date: Fri, 28 Nov 2025 15:09:22 +0800
Subject: arm64: dts: rockchip: add dma-coherent for pcie and gmac of RK3576

The RK3576 SoC employs ARM CCI for maintaining cache coherency
between the CPU cluster and high-speed peripherals including USB3,
SATA, GMAC, and PCIe controllers. While the USB3 and SATA controllers
were correctly marked as dma-coherent, the GMAC and PCIe nodes were
overlooked.

Without dma-coherent, the kernel falls back to software cache maintenance
for DMA operations, requiring explicit cache flushing and invalidating.
This adds significant overhead that degrades performance in high-throughput
workloads.

Add the missing dma-coherent properties to enable hardware coherency and
avoid unnecessary software cache management overhead.

Signed-off-by: Shawn Lin <shawn.lin@rock-chips.com>
---
 arch/arm64/boot/dts/rockchip/rk3576.dtsi | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/arm64/boot/dts/rockchip/rk3576.dtsi b/arch/arm64/boot/dts/rockchip/rk3576.dtsi
index 111111111111..222222222222 100644
--- a/arch/arm64/boot/dts/rockchip/rk3576.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3576.dtsi
@@ -709,6 +709,7 @@ pcie0: pcie@22000000 {
 			reset-names = "pwr", "pipe";
 			#address-cells = <3>;
 			#size-cells = <2>;
+			dma-coherent;
 			status = "disabled";
 
 			pcie0_intc: legacy-interrupt-controller {
@@ -763,6 +764,7 @@ pcie1: pcie@22400000 {
 			reset-names = "pwr", "pipe";
 			#address-cells = <3>;
 			#size-cells = <2>;
+			dma-coherent;
 			status = "disabled";
 
 			pcie1_intc: legacy-interrupt-controller {
@@ -1744,6 +1746,7 @@ gmac0: ethernet@2a220000 {
 			snps,mtl-rx-config = <&gmac0_mtl_rx_setup>;
 			snps,mtl-tx-config = <&gmac0_mtl_tx_setup>;
 			snps,tso;
+			dma-coherent;
 			status = "disabled";
 
 			mdio0: mdio {
@@ -1791,6 +1794,7 @@ gmac1: ethernet@2a230000 {
 			snps,mtl-rx-config = <&gmac1_mtl_rx_setup>;
 			snps,mtl-tx-config = <&gmac1_mtl_tx_setup>;
 			snps,tso;
+			dma-coherent;
 			status = "disabled";
 
 			mdio1: mdio {
-- 
Armbian

